batch/batch/driver/main.py

```diff
@@ -496,31 +534,33 @@ async def check(tx):
     jobs.cores_mcpu,
     (jobs.always_run OR NOT (jobs.cancelled OR batches.cancelled)) AS runnable,
     (NOT jobs.always_run AND (jobs.cancelled OR batches.cancelled)) AS cancelled,
-    batches.user
+    batches.user,
+    jobs.inst_coll
   FROM jobs
   INNER JOIN batches ON batches.id = jobs.batch_id
   WHERE batches.`state` = 'running'
   LOCK IN SHARE MODE) AS s
-GROUP BY user;
+GROUP BY user, inst_coll;
 ''')
-        computed_user_resources = {record['user']: record async for record in computed_user_resources}
+        computed_user_inst_coll_resources = {(record['user'], record['inst_coll']): record
+                                             async for record in computed_user_inst_coll_resources}
 
-        def user_get(d, u, f):
-            if u not in d:
+        def user_inst_coll_get(d, u_ic, f):
+            if u_ic not in d:
                 return 0
-            return d[u][f]
+            return d[u_ic][f]
 
         fields = ['n_running_jobs', 'running_cores_mcpu', 'n_ready_jobs', 'ready_cores_mcpu',
                   'n_cancelled_ready_jobs', 'n_cancelled_running_jobs']
-        users = set(user_resources.keys())
-        users.update(computed_user_resources.keys())
-        for u in users:
+        user_inst_colls = set(user_inst_coll_resources.keys())
+        user_inst_colls.update(computed_user_inst_coll_resources.keys())
+        for u_ic in user_inst_colls:
             for f in fields:
-                v = user_get(user_resources, u, f)
-                computed_v = user_get(user_resources, u, f)
+                v = user_inst_coll_get(user_inst_coll_resources, u_ic, f)
+                computed_v = user_inst_coll_get(user_inst_coll_resources, u_ic, f)
 
                 if v != computed_v:
-                    log.error(f'user_resources corrupt: user_resources[{u}][{f}] {v} != computed_user_resources[{u}][{f}] {computed_v}')
+                    log.error(f'user_inst_coll_resources corrupt: user_inst_coll_resources[{u_ic}][{f}] {v} != computed_user_inst_coll_resources[{u_ic}][{f}] {computed_v}')
```

Comment:
I find this python code kind of hard to follow. What do you think of just using the database to do this? It also saves a round trip and transferring unnecessary data.

We can avoid the `COALESCE` because there are no NULLs in these field (they're all `NOT NULL`). We don't need `IF` because booleans can be used as `1` (TRUE) and `0` (FALSE).
```python
async def check_incremental_loop(app, db):
    async def check():
        users_with_broken_resources = db.execute_and_fetchall('''
SELECT
  t.*,
  u.*
FROM
(
  SELECT user,
    SUM(state = 'Ready' AND runnable) AS actual_n_ready_jobs,
    SUM(cores_mcpu * (state = 'Ready' AND runnable)) AS actual_ready_cores_mcpu,
    SUM(state = 'Running' AND (NOT cancelled)) AS actual_n_running_jobs,
    SUM(cores_mcpu * (state = 'Running' AND (NOT cancelled))) AS actual_running_cores_mcpu,
    SUM(state = 'Ready' AND cancelled) AS actual_n_cancelled_ready_jobs,
    SUM(state = 'Running' AND cancelled) AS actual_n_cancelled_running_jobs
  FROM
  (
    SELECT batches.user, jobs.state, jobs.cores_mcpu,
      (jobs.always_run OR NOT (jobs.cancelled OR batches.cancelled)) AS runnable,
      (NOT jobs.always_run AND (jobs.cancelled OR batches.cancelled)) AS cancelled
    FROM batches
    INNER JOIN jobs ON batches.id = jobs.batch_id
    WHERE batches.`state` = 'running'
  ) as v
  group by user
) as t
INNER JOIN
(
  SELECT user, inst_coll,
    SUM(n_ready_jobs) as expected_n_ready_jobs,
    SUM(ready_cores_mcpu) as expected_ready_cores_mcpu,
    SUM(n_running_jobs) as expected_n_running_jobs,
    SUM(running_cores_mcpu) as expected_running_cores_mcpu,
    SUM(n_cancelled_ready_jobs) as expected_n_cancelled_ready_jobs,
    SUM(n_cancelled_running_jobs) as expected_n_cancelled_running_jobs
  from user_inst_coll_resources
  group by user, inst_coll
) as u
on t.user = u.user
WHERE actual_n_ready_jobs != expected_n_ready_jobs
   OR actual_ready_cores_mcpu != expected_ready_cores_mcpu
   OR actual_n_running_jobs != expected_n_running_jobs
   OR actual_running_cores_mcpu != expected_running_cores_mcpu
   OR actual_n_cancelled_ready_jobs != expected_n_cancelled_ready_jobs
   OR actual_n_cancelled_running_jobs != expected_n_cancelled_running_jobs
;
''')
        async for record in users_with_broken_resources:
            log.error(f'user_inst_coll_resources corrupt: {record}')

    while True:
        try:
            await check()  # pylint: disable=no-value-for-parameter
        except Exception as e:
            app['check_incremental_error'] = serialization.exception_to_dict(e)
            log.exception('while checking incremental')
        await asyncio.sleep(10)
```